### PR TITLE
Expose Quirks Mode information in the layout data and code

### PR DESCRIPTION
**Pull request by julienw**
This patch exposes the Quirks (NoQuirks/LimitedQuirks/Quirks) state to the layout subsystem.

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [ ] These changes fix #__ (github issue number if applicable).
Prelimary work for issue #11704.

- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because _____
(Waiting for guidance of where/which tests I could do here)

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/14430)

### CODE REVIEW

File: components/layout/fragment.rs [hunk 1]
```diff
@@ -3129,6 +3163,32 @@ bitflags! {
         const IS_INLINE_FLEX_ITEM = 0b0000_0001,
         /// Whether this fragment represents a child in a column flex container.
         const IS_BLOCK_FLEX_ITEM = 0b0000_0010,
+        /// Whether this fragment is part of a document in quirks mode.
+        const IS_QUIRKS_MODE = 0b0000_0100,
+        /// Which quirks mode we're in: 0 is quirks, 1 is limited quirks
+        const QUIRKS_MODE = 0b0000_1000,
+    }
```
Comment (julienw) [inline components/layout/fragment.rs]: here I decided to integrate it in the bitflags, because I thought fragments are everywhere and I didn't want to use more memory when not needed.

File: components/layout/generated_content.rs [hunk 2]
```diff
@@ -442,7 +442,8 @@ fn render_text(layout_context: &LayoutContext,
                                                              style.clone(),
                                                              style,
                                                              RestyleDamage::rebuild_and_reflow(),
-                                                             info));
+                                                             info,
+                                                             None));
```
Comment (julienw) [inline components/layout/generated_content.rs]: quirksmode information is actually useless here

File: components/layout/fragment.rs [hunk 3]
```diff
@@ -921,7 +922,34 @@ impl Fragment {
...
+            debug_id: DebugId::new(),
+            stacking_context_id: StackingContextId::new(0),
+        }
+    }
+
+    /// Constructs a new `Fragment` from a node, overriding its style
+    pub fn from_node_and_style<N: ThreadSafeLayoutNode>(node: &N,
+                                                        style: Arc<ServoComputedValues>,
+                                                        specific: SpecificFragmentInfo)
+                                                        -> Fragment {
```
Comment (julienw) [inline components/layout/fragment.rs]: I created this to simplify some of the existing code, but it's not _really_ necessary for this patch. Tell me if you'd prefer that I revert the changes around this constructor.

If you decide that we should keep it, maybe `new` could call it too and cut some code.

File: components/layout/fragment.rs [hunk 4]
```diff
@@ -921,7 +922,34 @@ impl Fragment {
...
+        let writing_mode = style.writing_mode;
+
+        let mut restyle_damage = node.restyle_damage();
+        restyle_damage.remove(RECONSTRUCT_FLOW);
```
Comment (julienw) [inline components/layout/fragment.rs]: If I read prperly, this single line is usually what is different where I changed the calls from `from_opaque_node_...` to this constructor. I don't know what are the consequences.

**Comment by emilio** [comment]
Isn't the Quirks mode a document-level thing? If so, shouldn't it go into `LayoutContext` instead of consuming memory all over the place?

**Comment by julienw** [comment]
Yes it is at the document-level; I'll check why I didn't go with this solution.

Note we don't consume more memory if we use the existing bitflags :)

**Comment by emilio** [comment]
Yes, but we'd use two flags that would be more helpful for other stuff :)

**Comment by julienw** [comment]
Yep, I'll look closer :)
Do you have some ideas about where tests could be added ?

**Comment by julienw** [comment]
So I looked closer and now I remember why I did it this way.

Ultimately we need the information in the flow objects (like InlineFlow or BlockFlow) for most of layout-related issues (I especially looked #11704 so far but I expect other issues to have the same requirements).

At first I only added the information in BaseFlow and modified the constructors for InlineFlow and BlockFlow (we don't even need to change LayoutContext for this as we can directly access the data in the document). InlineFlow was quite easy, but I saw that BlockFlow's constructors were used a lot with a fragment as argument. That's when I thought that having this information in a Fragment could be useful as well after all.

Now I understand how LayoutContext works: we create it once for every thread (with some shared data) and we pass it over as parameter to every call. Do you think it would be a better architecture ?

**Comment by emilio** [comment]
On Thu, Dec 01, 2016 at 07:10:18AM -0800, Julien Wajsberg wrote:
> Now I understand how LayoutContext works: we create it once for every
> thread (with some shared data) and we pass it over as parameter to
> every call. Do you think it would be a better architecture ?

Yes, I think so, makes sense to put in SharedLayoutContext anything that
is thread-independent and valid for the whole layout pass.

**Comment by julienw** [comment]
@emilio I just pushed 2 new commits to revert part of my previous changes and add the information in the shared layout context. Can you tell me what you think ?

File: components/script/layout_wrapper.rs [hunk 5]
```diff
@@ -908,6 +913,13 @@ impl<'ln> ThreadSafeLayoutNode for ServoThreadSafeLayoutNode<'ln> {
             self.get_jsmanaged().downcast::<Element>().unwrap().get_colspan()
         }
     }
+
+    #[inline]
+    fn quirks_mode(&self) -> QuirksMode {
```
Comment (emilio) [inline components/script/layout_wrapper.rs]: nit: This is not needed anymore, right?

File: components/layout_thread/lib.rs [hunk 6]
```diff
@@ -515,6 +521,7 @@ impl LayoutThread {
                 error_reporter: self.error_reporter.clone(),
                 local_context_creation_data: Mutex::new(local_style_context_creation_data),
                 timer: self.timer.clone(),
+                quirks_mode: self.quirks_mode.unwrap_or(QuirksMode::NoQuirks),
```
Comment (emilio) [inline components/layout_thread/lib.rs]: I think you should be able to `unwrap` here?

File: components/layout_thread/lib.rs [hunk 7]
```diff
@@ -228,6 +230,9 @@ pub struct LayoutThread {
     // Number of layout threads. This is copied from `util::opts`, but we'd
     // rather limit the dependency on that module here.
     layout_threads: usize,
+
+    /// Which quirks mode are we rendering the document in ?
```
Comment (emilio) [inline components/layout_thread/lib.rs]: nit: No space before `?`.

**Review by emilio — COMMENTED** [review]
Mostly nits, the rest looks sane to me :)

Also, please squash together the commits.

File: components/layout_thread/lib.rs (hunk 7 above)
Comment (julienw) [inline components/layout_thread/lib.rs]: my french typography ;)

File: components/layout_thread/lib.rs (hunk 6 above)
Comment (julienw) [inline components/layout_thread/lib.rs]: Looking only at this file, I wasn't sure `quirks_mode` would be always already populated when eg `repaint` or `tick_animations` or others would be called. I think it's not a big deal to have a default value here, what do you think ?

File: components/script/layout_wrapper.rs (hunk 5 above)
Comment (julienw) [inline components/script/layout_wrapper.rs]: yeah, I think you're right. If we ever need it later, we can just add it at that time.

**Comment by julienw** [comment]
I fixed your comments except the last one, I squashed the commits.

I'm not sure of what gives the test issue on Travis. I reproduce also locally, but not on master, so it should be something in this patch. But I don't know how to get more information.

File: components/layout_thread/lib.rs (hunk 6 above)
Comment (emilio) [inline components/layout_thread/lib.rs]: I think it'd be wrong to lay out a quirks mode document in no-quirks mode, then laying it out in quirks mode, overall taking into account incremental reflows.

Comment (julienw) [inline components/layout_thread/lib.rs]: But just unwrapping would panic the thread in the same case; is it wishable  ?

Comment (emilio) [inline components/layout_thread/lib.rs]: I think we do. I don't think we can layout anything without a document anyway. Indeed the sanest stuff would be making `build_shared_layout_context` taking the document directly, and don't store the quirks mode in the script thread at all.

File: components/style/Cargo.toml [hunk 8]
```diff
@@ -30,6 +30,7 @@ euclid = "0.10.1"
 fnv = "1.0"
 heapsize = "0.3.0"
 heapsize_derive = {version = "0.1", optional = true}
+html5ever = {version = "0.10.2", features = ["heap_size", "unstable"]}
```
Comment (SimonSapin) [inline components/style/Cargo.toml]: This is a fairly large dependency to add to Stylo, just to use a small enum. And it is also what causes `build-geckolib` to fail because that uses an older (stable channel) Rust compiler that doesn’t know about new-style procedural macros.

Instead I recommend duplicating the definition of the `QuirksMode` enum into `components/style/context.rs`. For consistency, the rest of Servo should be changed to use that enum instead of `html5ever::tree_builder::QuirksMode`. Namely, these files:

```
components/script/dom/node.rs
components/script/dom/document.rs
components/script/dom/bindings/trace.rs
components/script/dom/servoparser/html.rs
```

The latter has a `set_quirks_mode` method that takes a `html5ever::tree_builder::QuirksMode` and will need to do the conversion to `style::context::QuirksMode`.

File: Cargo.lock [hunk 9]
```diff
@@ -223,6 +223,13 @@ version = "0.1.17"
 source = "git+https://github.com/browserhtml/browserhtml?branch=crate#aafbb0996b02d1fadd4713c96e3d22b542f175a1"
 
 [[package]]
+name = "build-apk"
```
Comment (Ms2ger) [inline Cargo.lock]: Revert

**Comment by emilio** [comment]
Does this info need to be in the `SharedStyleContext`? Shouldn't it be better for it to be in the `SharedLayoutContext` if only layout is going to use it?

**Comment by julienw** [comment]
> Does this info need to be in the SharedStyleContext? Shouldn't it be better for it to be in the SharedLayoutContext if only layout is going to use it?

Yeah that was also a question I asked to @SimonSapin :) He said that some of the things in the quirks mode are style-related, for example https://quirks.spec.whatwg.org/#the-tables-inherit-color-from-body-quirk

File: Cargo.lock (hunk 9 above)
Comment (julienw) [inline Cargo.lock]: sure, I don't know where this comes from.

Comment (SimonSapin) [inline Cargo.lock]: This was a bug unrelated to this PR, it’s fixed on master now. This should disappear if you rebase.

File: ports/geckolib/glue.rs [hunk 10]
```diff
@@ -115,6 +115,8 @@ fn create_shared_context(mut per_doc_data: &mut AtomicRefMut<PerDocumentStyleDat
         error_reporter: Box::new(StdoutErrorReporter),
         local_context_creation_data: Mutex::new(local_context_data),
         timer: Timer::new(),
+        // FIXME Find the real QuirksMode information for this document
```
Comment (julienw) [inline ports/geckolib/glue.rs]: do you want a bugzilla bug number for this ?

**Comment by julienw** [comment]
Green tests !

**Comment by julienw** [comment]
> ☔️ The latest upstream changes (presumably #13489) made this pull request unmergeable. Please resolve the merge conflicts. 

done

**Comment by emilio** [comment]
Looks good, thanks!

@bors-servo r+

**Comment by jdm** [comment]
@bors-servo: delegate+
If you rebase again, you can merge it immediately with r=emilio.

**Comment by julienw** [comment]
Just realized how I should add r=emilio :) but afk right now so this will wait a few more hours.

**Comment by julienw** [comment]
thanks @kichjang !

**Comment by emilio** [comment]
Thank you for the patch @julienw! :)
